### DISCUSSION TITLE
Documentation... Like, *actual* documentation.

### DISCUSSION THREAD

**Discussion by Derjyn** [HolterPhylo/AMUMSS#30] [General]
Digging into modding NMS, AMUMSS seems to be pretty amazing save for one thing: the amazing jumble of what we might call documentation.

Seeing as how this is the primary repository, one would think the wiki or core readme would have the initial speck of a breadcrumb trail for getting started with developing mods using LUA. However, take a quick peek at the [wiki](https://github.com/HolterPhylo/AMUMSS/wiki) here, or the [README](https://github.com/HolterPhylo/AMUMSS/blob/main/README.md)... It's a lot to take in, of course.

Thankfully there is a link to Discord- that actually links to a Nexus page, that then links to a Discord, that is a pretty awesome hangout for NMS community denizens... but lacks any to-the-point developer reference. Grumpy 40 year olds wearing an arrogant script kiddie costume, that make annoyed noises at anyone asking about modding don't count as a developer reference.

This repo is years old, so it's baffling that there isn't a hello world and roadmap for developers to follow. Here we have some scripts:

https://github.com/MetaIdea/nms-amumss-lua-mod-script-collection

That's pretty much the gold pile I've been able to find so far. Where are the best practices? How everything works? Is there any sort of template (that doesn't have a caveat of "_don't use this!_") kicking around?

When this sort of developer energy is the hidden knowledge cliche, it stunts the growth of quality and creativity. Thankfully we can be stubborn and fumble around in the dark until we make something work, _but that doesn't need to be so_. 

* Quick how to get started/hello world
* Best practices
* Common pitfalls
* Under the hood
* Contribution needs/clear roadmap

I hope I just haven't found it yet, but if it's taken this long to come up empty handed, something is lacking. Other links related to AMUMSS for reference of what developers are working with:

https://www.nexusmods.com/nomanssky/mods/2626
https://stepmodifications.org/wiki/NoMansSky:AMUMSS_How-tos
https://stepmodifications.org/wiki/NoMansSky:AMUMSS_Guide/ReadMe
https://stepmodifications.org/wiki/NoMansSky:AMUMSS_Guide

**Comment by HolterPhylo** [comment]
I am sorry that you find all this confusing.

I hope you come to the Discord and ask questions.  
[https://discord.gg/HFjnmnwe67](https://discord.gg/HFjnmnwe67)

This here is just a place to house releases and updates to AMUMSS (because Nexus was always blocking for AV scanning and auto-updating was not possible) AND send you where people with knowledge about NMS and AMUMSS are, nothing else.

I corrected the links, I hope.

- Quick how to get started/hello world
   > Once installed, you get a link to _README-AMUMSS_Script_Rules which is the bible for all AMUMSS commands
- Best practices
   > There are a lot of examples scripts and many ways to do the same thing, so explore and read
- Common pitfalls
   > After the installation hurdle :)
   > The most common pitfalls is that people do not read instructions
- Under the hood
   > Not sure what you mean by that, everything is open source
- Contribution needs/clear roadmap
   > Contributions are welcome.  I clearly welcome suggestions.  Discord room is the place to do that.
   > Roadmap: While maintaining backward compatibility as must as possible, add new features from suggestions and perceived needs

**Comment by Derjyn** [comment]
Awesome. That Discord is a different spot than what was linked to in other places, and is certainly more developer-focused.

"Under the hood" or "behind the curtain" is a phrase that means... well, talking about how the technicals work. Commonly a "this happens here, this thing does that, this happens there, etcetera etcetera". Under the hood of a vehicle. Explaining how the engine runs. It's a phrase that's been around for longer than I've been alive, and that's a hot minute.

That __README-AMUMSS_Script_Rules_ file is excellent reference, but I wouldn't call it a bible. It's more akin to an alchemists reference, expecting that the reader is already a master alchemist. What is missing, is the apprentice, journeyman, or hobbyist documentation that leads up to that level of reference.

I have the C++ primer sitting on a bookshelf next to me - it's about 1,000 pages. C++ is an advanced compiled language, and this 1000 page book is just the primer... Even that has a beginner hand-holding section to get readers on their way to the more advanced inner dark magick.

My point is, I think we (meaning the poor naïve beginners) modders wouldn't complain if there was a starting point that made a logical progression into the current level of documentation and resources. Currently what we have assumes everyone is already a seasoned pro with NMS modding, the LUA scripting language, and whatever the heck AMUMSS is. Essentially there is a disconnect.

Totally get it if you don't have time, seeing as how you're focused on AMUMSS, and documentation and reference is a heck of a task in it's own right. Piecing together a pathway of getting started to more advanced modding should be possible between the information on the Discord and the script rules (and other sparse) documentation. Painful and clumsy, but that's what we developers should be equipped to handle!

To make it clear, I don't personally find this all confusing, and I didn't fail to RTFM. This level of documentation on a tool TJW (_that just works_) isn't a foreign experience. Having a basic, short and to-the-point tutorial on getting a basic mod up and running would be awesome. [This](https://github.com/MetaIdea/nms-amumss-lua-mod-script-collection/blob/b2b872de6fca555536d97163977935e7fed3eb68/AMUMSS-ModLuaLearningCollection/These%20scripts%20ARE%20for%20LEARNING%20ONLY/LearningExamples/TEMPLATE.lua) is not beginner friendly.

Keep up the great work! Don't take anything I'm firing off here as harsh criticism. It's more of a dialogue than anything, and meant in a friendly and nerdy/technical tone.

**Comment by CodeX-SC** [comment]
Derjyn
With AMUMSS we have to take into mind that when it was made quite a few years ago , im going to assume it was made in such a way that someone with no exp in lua or possibly any form of programming exp - would be able to use it as it gave simple references of what it used - wasnt really any lua code in the scripts. Come forward to the present - quite a lot of things have been improved upon and slight changes made to the non lua code we can use , AMUMSS as it states will also accept lua code in the script as well and this is where things can seem to go from short and easy to omg i have to learn lua now to use it.
Theres a lot of nms modders who have learnt programming and know lua and thus use lua code etc . I tend to look at their scripts and their lua code and work out what its doing etc - plus i can refer to lua manual and script rules etc . Im no programmer and im slowly learning python which helps as i can recognise similarites between the 2 languages.
yes i will agree that even tho theres examples etc , some stuff becomes not beginner friendly as the example doesnt give enough details .
I may suggest on discord - if theres any authors who use lua code could they provide some reference scripts to explain whats doing what etc
